TheWidlarzGroup/react-native-video
# Feature/android pip

## Provide an example of how to test the change
In order to test this change, you should follow all new steps listed in readme under picureInPicture props:

* Add following to the AndroidManifest.xml -> MainActivity:
```
  android:configChanges="keyboard|keyboardHidden|orientation|screenSize|screenLayout"
  android:resizeableActivity="true"
  android:supportsPictureInPicture="true"
```
* [Enable PIP feature for you app](https://support.google.com/youtube/answer/7552722?co=GENIE.Platform%3DAndroid&hl=en) in your device settings
* Add following methods to MainActivity:
```
    // To let JS knows about PIP mode change. 
    override fun onPictureInPictureModeChanged(isInPictureInPictureMode: Boolean, newConfig: Configuration?) {
       super.onPictureInPictureModeChanged(isInPictureInPictureMode, newConfig)
       val intent = Intent("onPictureInPictureModeChanged")
       intent.putExtra("isInPictureInPictureMode", isInPictureInPictureMode)
       this.sendBroadcast(intent)
    }
    
    // To trigger PIP mode by pressing `Home` or `Recent` buttons
    override fun onUserLeaveHint() {
       val intent = Intent("onUserLeaveHint")
       this.sendBroadcast(intent)
       super.onUserLeaveHint()
    }
        
    // To close and kill the app when closing PIP window.
    override fun onStop() {
       super.onStop()
       finishAndRemoveTask()
    }
``` 

#### Afterwards, set `pictureInPicture` prop, or press Home/Recent HW button.

### Describe the changes
So far PIP mode has been supported only on iPad. From now on, it should be working also on all Android Phones and Tablets running Oreo at least. This implementation includes:

- Ability to toggle PIP mode by setting `pictureInPicture` prop.
- `onPictureInPictureStatusChanged` prop callback returning PIP state same as iOS does.
- Ability to turn on PIP mode by pressing Home/Recent HW button.
- Previous / Pause/Play / Next PIP window buttons.
- Implementation of closing and killing the entire app when closing PIP window.

![android-PIP](https://user-images.githubusercontent.com/31282181/65879688-a6846100-e390-11e9-88d9-13a8343cfe9f.gif)
